### PR TITLE
parser: (partially) fix float point parsing with leading dot

### LINKED ISSUE #3954: cannot parse float with leading dot && non trailing space
### &, |

TiDB:

```
mysql> select .78&&0;
ERROR 1105 (HY000): line 0 column 8 near "78&&0" (total length 13)
```

MySQL:

```
mysql> select .78&&0;
+--------+
| .78&&0 |
+--------+
|      0 |
+--------+
1 row in set (0.01 sec)
```

### +, -, *, /

TiDB:

```
mysql> select .78+4;
ERROR 1105 (HY000): line 0 column 8 near "78+4" (total length 12)
```

MySQL:

```
mysql> select .78+4;
+-------+
| .78+4 |
+-------+
|  4.78 |
+-------+
1 row in set (0.00 sec)
```

### ,

See https://github.com/pingcap/tidb/issues/3889

### PR DESCRIPTION
**Pull request by breeswish**
Fix #3954 and #3889

Previous related commits: https://github.com/pingcap/tidb/commit/5917a784be944903c12f2ce42d57665dbc608904

Note: This PR does not fully solves compatibility issues, see #3963

@tiancaiamao PTAL, thanks

### CODE REVIEW

File: parser/lexer_test.go [hunk 1]
```diff
@@ -116,8 +116,17 @@ func (s *testLexerSuite) TestLiteral(c *C) {
...
+		{".1_t_1_x", int('.')}, // `.`, `1_t_1_x`
+		{".1$23", int('.')},    // `.`, `1$23`
+		{".1a23", int('.')},    // `.`, `1a23`
+		{".1C23", int('.')},    // `.`, `1C23`
+		{".1\u0081", int('.')}, // `.`, `1\u0081`
+		{".1\uff34", int('.')}, // `.`, `1\uff34`
+		{".123", decLit},       // `.123`
+		{".1*23", decLit},      // `.1`, `*`, `23`
+		{".1,23", decLit},      // `.1`, `,`, `23`
+		{".1 23", decLit},      // `.1`, `23`
```
Comment (tiancaiamao) [inline parser/lexer_test.go]: add a test `.1e23$23` , rest looks good to me.

Comment (breeswish) [inline parser/lexer_test.go]: @tiancaiamao Done

**Comment by tiancaiamao** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**